ksqldb-common/src/main/java/io/confluent/ksql/schema/ksql/SystemColumns.java

```diff
@@ -134,28 +143,58 @@ public static boolean isSystemColumn(final ColumnName columnName, final KsqlConf
     return isSystemColumn(columnName, getPseudoColumnVersionFromConfig(ksqlConfig));
   }
 
-  @SuppressFBWarnings(
-      value = "MS_EXPOSE_REP",
-      justification = "SYSTEM_COLUMN_NAMES is ImmutableSet"
-  )
-  private static Set<ColumnName> systemColumnNames(final int pseudoColumnVersion) {
-    return SYSTEM_COLUMN_NAMES_BY_VERSION.get(pseudoColumnVersion);
+  public static Set<ColumnName> systemColumnNames(final int pseudoColumnVersion) {
+    return Stream.concat(
+        WINDOW_BOUNDS_COLUMN_NAMES.stream(),
+        pseudoColumns
+            .stream()
+            .filter(col -> col.version <= pseudoColumnVersion)
+            .map(col -> col.name)
+    )
+        .collect(Collectors.toSet());
   }
 
   public static boolean mustBeMaterializedForTableJoins(final ColumnName columnName) {
-    return MUST_BE_MATERIALIZED_FOR_TABLE_JOINS.contains(columnName);
+    return pseudoColumns
+        .stream()
+        .filter(col -> col.name.equals(columnName))
+        .findFirst().orElseThrow(IllegalArgumentException::new)
+        .mustBeMaterializedForTableJoins;
   }
 
-  private static Set<ColumnName> buildColumns(final int pseudoColumnVersion) {
-    return ImmutableSet.<ColumnName>builder()
-        .addAll(pseudoColumnNames(pseudoColumnVersion))
-        .addAll(WINDOW_BOUNDS_COLUMN_NAMES)
-        .build();
-  }
-
-  private static int getPseudoColumnVersionFromConfig(final KsqlConfig ksqlConfig) {
+  public static int getPseudoColumnVersionFromConfig(final KsqlConfig ksqlConfig) {
     return ksqlConfig.getBoolean(KsqlConfig.KSQL_ROWPARTITION_ROWOFFSET_ENABLED)
         ? CURRENT_PSEUDOCOLUMN_VERSION_NUMBER
         : LEGACY_PSEUDOCOLUMN_VERSION_NUMBER;
   }
+
+  @Immutable
+  private static final class PseudoColumn {
+
+    final ColumnName name;
+    final SqlType type;
+    final int version;
+    final boolean mustBeMaterializedForTableJoins;
+
+    private PseudoColumn(
+        final ColumnName name,
+        final SqlType type,
+        final int version,
+        final boolean mustBeMaterializedForTableJoins
+    ) {
+      this.name = name;
+      this.type = type;
+      this.version = version;
+      this.mustBeMaterializedForTableJoins = mustBeMaterializedForTableJoins;
+    }
+
+    private static PseudoColumn of(
```

Comment:
Out of curiosity, what's the benefit of having a static creator method here?

Reply:
Just readability. If you think it hurts readability or doesn't contribute, I don't feel strongly about it either way.

Reply:
I don't think it contributes but I'm also not opposed to leaving it, was just curious what your motivation was.